watcher/local/statespubsub.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/pkg/sync/atomic"
+	"perun.network/go-perun/watcher"
+)
+
+var _ watcher.StatesPub = &statesPubSub{}
+
+// errAlreadyClosed indicates the pub-sub has already been closed.
+var errAlreadyClosed error = errors.New("already closed")
+
+const statesPubSubBufferSize = 10
+
+type (
+	statesPubSub struct {
+		closed atomic.Bool
+		pipe   chan channel.Transaction
+	}
+
+	// statesSub is used by the watcher to receive transactions.
+	statesSub interface {
+		statesStream() <-chan channel.Transaction
+		error() error
+		close() error
+	}
+)
+
+func newStatesPubSub() *statesPubSub {
+	return &statesPubSub{
+		pipe: make(chan channel.Transaction, statesPubSubBufferSize),
+	}
+}
+
+// Publish publishes the given state to all the subscribers.
+//
+// Each time when a transaction is published, watcher will treat it as the
+// latest transaction without any validation. It is the responsibility of the
+// client to publish transactions in correct order and ensure they are the
+// valid.
+//
+// Context is added as this method implements watcher.StatesPub interface.
+// However, is is not used as there no network communications involved in local
+// watcher.
+//
+// Returns nil if the state is published. Panics if the subscriptions is
```

Comment:
```suggestion
// Returns nil if the state is published. Panics if the subscription is
```

Reply:
Can we change the implementation so that it throws an error instead of panicing if the channel is already closed?

Reply:
We could do that, but it would come at a cost.

1. We need to use a variable + mutex to co-ordinate the publishing and closing. Because these aspects are done by two different components and should be concurrency-safe.
2. Then, each time we publish, we need to acquire mutex lock, check if the channel is closed and then publish.

I think, it is not worth introducing these complexities in this case, because the close will be called only after client calls stop watching. And it is fair to expect that, the client does not `Publish` after it calls `StopWatching.

Reply:
Reading your comment, it seems you would not have implemented my suggestion.

However, reading the code, it looks like you did.

Am I correct? What is the current state?

Reply:
👌